### PR TITLE
feat(centos-8): added centos 8 support

### PR DESCRIPTION
**Pull request by aboe76**
moved RedHat/CentOS to osmap and osfingermap to accomodate changes in
package provider versionlock

Do not merge yet, waiting on salt bug: saltstack/salt#54798
Centos 8 uses the wrong package provider, should be dnf but it's using yum
which results in an error with yum-plugin-versionlock, which isn't available on centos-8.

### PR progress checklist (to be filled in by reviewers)

- [ ] Changes to documentation are appropriate (or tick if not required)
- [ ] Changes to tests are appropriate (or tick if not required)
- [ ] Reviews completed

---

#### Primary type

- [ ] `[build]`    Changes related to the build system
- [ ] `[chore]`    Changes to the build process or auxiliary tools and libraries such as documentation generation
- [ ] `[ci]`       Changes to the continuous integration configuration
- [x] `[feat]`     A new feature
- [ ] `[fix]`      A bug fix
- [ ] `[perf]`     A code change that improves performance
- [ ] `[refactor]` A code change that neither fixes a bug nor adds a feature
- [ ] `[revert]`   A change used to revert a previous commit
- [ ] `[style]`    Changes that do not affect the meaning of the code (white-space, formatting, missing semi-colons, etc.)

#### Secondary type

- [ ] `[docs]`     Documentation changes
- [ ] `[test]`     Adding missing or correcting existing tests

### Does this PR introduce a `BREAKING CHANGE`?

No.

### Documentation checklist

- [ ] Updated the `README` (e.g. `Available states`).
- [ ] Updated `pillar.example`.

### Testing checklist

- [ ] Included in Kitchen (i.e. under `state_top`).
- [x] Covered by new/existing tests (e.g. InSpec, Serverspec, etc.).
- [x] Updated the relevant test pillar.

### CODE REVIEW

**Comment by myii** [comment]
Thanks for re-opening this @aboe76.  The PR title will be perfect as the commit message, so that will allow the tests to run.

```diff
-Add centos 8 to osfingermap
+feat(centos-8): added centos 8 support
```

The commit body can stay the same, that shouldn't cause a problem.

**Comment by myii** [comment]
@aboe76 Thanks for updating this.  Unfortunately, we haven't enabled any of the `centos-8` images for testing yet!  I'll try that right now in my fork and let you know what the results are.

**Comment by myii** [comment]
@aboe76 OK, we've got a few failing states (`Succeeded: 13 (changed=5); Failed:     8`) for all of the instances:

* https://gitlab.com/myii/packages-formula/-/pipelines/245878109

Would you mind enabling `centos-centos-8-master-py3` in this PR so that we can resolve there failures as well?

**Comment by myii** [comment]
Hmm, the CI run for the very last commit (572665a) isn't showing up here but it was pushed through to GitLab, it seems:

* https://gitlab.com/saltstack-formulas/packages-formula/-/pipelines/245925810
  - Unfortunately, both `centos-7` and `centos-8` are failing.

This is the first time this has happened, so I'll ask the `coqbot` team about this.

**Comment by myii** [comment]
Sent a question to them: https://coq.zulipchat.com/#narrow/stream/243318-coqbot-devs.20.26.20users/topic/GitHub.20reporting.20failing.20after.20numerous.20pushes.20to.20a.20PR/near/223774255.

**Comment by aboe76** [comment]
strange, it has to do with 'yum-plugin-versionlock' but I need to specify this for the following machines:
RedHat/Centos 6/7 : yum-plugin-versionlock
RedHat/Centos 8: dnf-plugin-versionlock

So that's why there is a osfingermap file, but I need to know what the RedHat osfinger looks like to complete it.

**Comment by myii** [comment]
@aboe76 Strange, the GitLab report is showing up now.  Did you do anything?

Anyway, the issue seems to be with the test pillar file: https://github.com/saltstack-formulas/packages-formula/blob/f893e2af43bf547d64cbbb3b06c27bef767c65ff/test/salt/pillar/redhat.sls#L17-L20.

Probably can remove both of those, so that they don't interfere with the `pkgs:required:pkgs` defined in the `os*.yaml` files.

The other failure seems to be with the `held` packages.

**Comment by aboe76** [comment]
> @aboe76 Strange, the GitLab report is showing up now. Did you do anything?
> 
> Anyway, the issue seems to be with the test pillar file:
> 
> https://github.com/saltstack-formulas/packages-formula/blob/f893e2af43bf547d64cbbb3b06c27bef767c65ff/test/salt/pillar/redhat.sls#L17-L20
> 
> .
> Probably can remove both of those, so that they don't interfere with the `pkgs:required:pkgs` defined in the `os*.yaml` files.
> 
> The other failure seems to be with the `held` packages.

Let's test

**Comment by Yoda-BZH** [comment]
Hello,

Any news ?

Thank you !

**Comment by aboe76** [comment]
I still have issues, in the pipeline testing with one single pillar on "redhat" machines...

**Comment by aboe76** [comment]
@myii can you take a look why centos8 takes so long?

**Comment by myii** [comment]
> @myii can you take a look why centos8 takes so long?

@aboe76 It appeared to have been cancelled so I restarted it and now it's all working except one of the tests:

* https://gitlab.com/saltstack-formulas/packages-formula/-/jobs/1013667577

The problem is here:

```rb
# Prepare platform "finger"
platform_finger = platform[:finger].split('.').first.to_s
```

There isn't a `platform[:finger]` provided by InSpec.  We've actually got a shared library available that provides that and then you use:

```diff
-platform_finger = platform[:finger].split('.').first.to_s
+platform_finger = system.platform[:finger].split('.').first.to_s
```

So you'll need to copy this entire directory over from the `template-formula` as well:

* https://github.com/saltstack-formulas/template-formula/tree/master/test/integration/share

Once that's confirmed working, there are just a couple of other minor issues I'd like to discuss when we rebase this PR.

**Comment by aboe76** [comment]
@myii it's working with the 'share' folder

**Comment by aboe76** [comment]
this also fixes: #66

File: .gitlab-ci.yml [hunk 1]
```diff
@@ -126,7 +126,7 @@ rubocop:
 debian-debian-10-master-py3: {extends: '.test_instance'}
 # ubuntu-ubuntu-2004-master-py3: {extends: '.test_instance'}
 ubuntu-ubuntu-1804-master-py3: {extends: '.test_instance'}
-# centos-centos-8-master-py3: {extends: '.test_instance'}
+centos8-centos-8-master-py3: {extends: '.test_instance'}
```
Comment (myii) [inline .gitlab-ci.yml:129]:
```suggestion
redhat8-centos-8-master-py3: {extends: '.test_instance'}
```

See below.

File: kitchen.yml [hunk 2]
```diff
@@ -240,10 +240,32 @@ suites:
     verifier:
       inspec_tests:
         - path: test/integration/default
-  - name: centos
+  - name: centos8
```
Comment (myii) [inline kitchen.yml:243]:
```suggestion
  - name: redhat8
```

Will be required for `oraclelinux-8` as well, so suggest renaming this accordingly (also matches the test pillar name below).

File: packages/osfingermap.yaml [hunk 3]
```diff
@@ -0,0 +1,23 @@
+# -*- coding: utf-8 -*-
+# vim: ft=yaml
+---
+CentOS Linux-6:
```
Comment (myii) [inline packages/osfingermap.yaml:4]: This is the actual `osfinger` for this platform:

```suggestion
CentOS-6:
```

**Review by myii — CHANGES_REQUESTED** [review]
Generally fine, @aboe76 -- just a couple of fixes suggested.

In fact, this commit contains everything, including adding a test run for `oraclelinux-8`:

* https://github.com/myii/packages-formula/commit/d74462a65f4ddfddd62a11abc812dd14ddb595d3

Here's the corresponding CI run:

* https://gitlab.com/myii/packages-formula/-/pipelines/253493343

I'm not sure if I'm able to push that here directly because this PR was sent through from your `master` branch.

**Comment by aboe76** [comment]
@myii are you happy with RedHat, Centos and Oracle linux 8 support?

File: .gitlab-ci.yml [hunk 4]
```diff
@@ -126,7 +126,8 @@ rubocop:
 debian-debian-10-master-py3: {extends: '.test_instance'}
 # ubuntu-ubuntu-2004-master-py3: {extends: '.test_instance'}
 ubuntu-ubuntu-1804-master-py3: {extends: '.test_instance'}
-# centos-centos-8-master-py3: {extends: '.test_instance'}
+redhat8-centos-8-master-py3: {extends: '.test_instance'}
+redhat8-oraclelinux-8-master-py3: {extends: '.test_instance'}
```
Comment (myii) [inline .gitlab-ci.yml:130]: This line will end up getting moved lower down in the file once I update the `ssf-formula` with this PR but nothing to worry about (I'll just push the commit on top after this PR is merged).  Just for your info, @aboe76.

**Review by myii — APPROVED** [review]
Nice work, @aboe76!  Thanks for all of the patience it took.

**Comment by Yoda-BZH** [comment]
Apparently the correction for `Red Hat Enterprise Linux-8` (osfinger) is not complete.

I still have :
```
local:
----------
          ID: pkg_req_pkgs
    Function: pkg.installed
      Result: False
     Comment: Attempt 1: Returned a result of "False", with the following comment: "The following packages failed to install/update: yum-plugin-versionlock"
              Attempt 2: Returned a result of "False", with the following comment: "The following packages failed to install/update: yum-plugin-versionlock"
              The following packages failed to install/update: yum-plugin-versionlock
     Started: 15:15:05.139721
    Duration: 154690.206 ms
     Changes:   
----------
```

Redhat 8 Enterprise uses dnf, not yum

**Comment by myii** [comment]
> Apparently the correction for `Red Hat Enterprise Linux-8` (osfinger) is not complete.

@Yoda-BZH Yes, unfortunately we don't have CI testing for that at the moment.  If I recall correctly, the `salt-bootstrap` doesn't complete the installation successfully for it.

Please feel free to submit a PR and we can get that merged pretty quickly.

**Pull request merged**